src/unistd.rs

```diff
@@ -2392,3 +2397,425 @@ pub fn access<P: ?Sized + NixPath>(path: &P, amode: AccessFlags) -> Result<()> {
+///
+/// The buffer size variants should not be needed 99% of the time; the default buffer size of 1024
+/// should be more than adequate. Only use the buffer size variants if you receive an ERANGE error
+/// without them, or you know that you will be likely to.
+///
+/// # Examples
+///
+/// ```
+/// use nix::unistd::{User, UserQuery, Queryable};
+/// let find = String::from("root");
+/// // Returns an Option<Result<User>>, thus the double unwrap.
+/// let res = User::query( UserQuery::Name(find.clone()) ).unwrap().unwrap();
+/// assert!(res.name == find);
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum UserQuery {
+    /// Get a user by UID.
+    ///
+    /// Internally, this function calls
+    /// [getpwuid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Uid(Uid),
+    /// Get a user by name.
+    ///
+    /// Internally, this function calls
+    /// [getpwnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    UidWithBufsize(Uid, usize),
+    NameWithBufsize(String, usize)
+}
+
+/// Query a group.
+///
+/// The buffer size variants should not be needed 99% of the time; the default buffer size of 1024
+/// should be more than adequate. Only use the buffer size variants if you receive an ERANGE error
+/// without them, or you know that you will be likely to.
+///
+/// # Examples
+///
+/// ```ignore
+/// use nix::unistd::{Group, GroupQuery};
+/// // On many systems there's no `root` group; on FreeBSD for example it's known as `wheel`
+/// let find = String::from("root");
+/// // Returns an Option<Result<Group>>, thus the double unwrap. Will panic if there's no `root`
+/// // group.
+/// let res = Group::query( GroupQuery::Name(find.clone()) ).unwrap().unwrap();
+/// assert!(res.name == find);
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum GroupQuery {
+    /// Get a group by GID.
+    ///
+    /// Internally, this function calls
+    /// [getgrgid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Gid(Gid),
+    /// Get a group by name.
+    ///
+    /// Internally, this function calls
+    /// [getgrnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    /// Get a group by GID, explicitly specifying the buffer size.
+    GidWithBufsize(Gid, usize),
+    /// Get a group by name, explicitly specifying the buffer size.
+    NameWithBufsize(String, usize)
+}
+
+pub trait Queryable<Q> where Self: ::std::marker::Sized {
+    fn query(q: Q) -> Option<Result<Self>>;
+}
+
+/// Default buffer size for system user and group querying functions
+const PWGRP_BUFSIZE: usize = 1024;
+
+impl Queryable<UserQuery> for User {
+    fn query(q: UserQuery) -> Option<Result<Self>> {
+        let mut cbuf = match q {
+            UserQuery::UidWithBufsize(_, size) |
+            UserQuery::NameWithBufsize(_, size) => vec![0 as c_char; size],
```

Comment:
Instead of zero-initializing, this should be uninitialized, a little like this:
```rust
let mut cbuf = Vec::with_capacity(size);
let error = libc::getpwuid_r(..., ..., cbuf.as_mut_ptr(), cbuf.len(), ...);
```
Ordinarily you'd follow up with a `cbuf.set_len`.  But in this case, since the vector is being accessed solely by C code, there's no need.